Sources/App/Commands/Analyzer.swift

```diff
@@ -24,48 +24,164 @@ struct AnalyzerCommand: Command {
+    do {
+        return try pullOrClone(application: application, package: package)
+            .map { .success($0) }
+            .flatMapErrorThrowing { .failure($0) }
+    } catch {
+        return application.eventLoopGroup.next().makeSucceededFuture(.failure(error))
     }
 }
 
 
 func pullOrClone(application: Application, package: Package) throws -> EventLoopFuture<Package> {
-    guard let basename = package.localCacheDirectory else {
+    guard let path = Current.fileManager.cacheDirectoryPath(for: package) else {
         throw AppError.invalidPackageUrl(package.id, package.url)
     }
-    let path = application.directory.checkouts + "/" + basename
     return application.threadPool.runIfActive(eventLoop: application.eventLoopGroup.next()) {
-        if FileManager.default.fileExists(atPath: path) {
+        if Current.fileManager.fileExists(atPath: path) {
             application.logger.info("pulling \(package.url) in \(path)")
-            try shellOut(to: .gitPull(), at: path)
+            try Current.shell.run(command: .gitPull(), at: path)
         } else {
             application.logger.info("cloning \(package.url) to \(path)")
-            try shellOut(to: .gitClone(url: URL(string: package.url)!, to: path))
+            try Current.shell.run(command: .gitClone(url: URL(string: package.url)!, to: path))
         }
         return package
     }
 }
+
+
+/// Wrapper around _reconcileVersions to create a non-throwing version (mainly to ensure
+/// that failed futures don't slip through and break the pipeline).
+func reconcileVersions(application: Application, result: Result<Package, Error>) -> EventLoopFuture<(Package, [Version])> {
+    do {
+        let pkg = try result.get()
+        return try _reconcileVersions(application: application, package: pkg)
+            .map { (pkg, $0) }
+    } catch {
+        return application.eventLoopGroup.next().makeFailedFuture(error)
+    }
+}
+
+
+func _reconcileVersions(application: Application, package: Package) throws -> EventLoopFuture<[Version]> {
+    // fetch tags
+    guard let path = Current.fileManager.cacheDirectoryPath(for: package) else {
+        throw AppError.invalidPackageUrl(package.id, package.url)
+    }
+    guard let pkgId = package.id else {
+        throw AppError.genericError(nil, "PANIC: package id nil for package \(package.url)")
+    }
+    let tags: EventLoopFuture<[String]> = application.threadPool.runIfActive(eventLoop: application.eventLoopGroup.next()) {
+        application.logger.info("listing tags for package \(package.url)")
+        let tags = try Current.shell.run(command: .init(string: "git tag"), at: path)
+        return tags.split(separator: "\n").map(String.init)
+    }
+    // FIXME: also save version for default branch (currently only looking at tags)
+
+    // TODO: sas 2020-05-02: is it necessary to reconcile versions or is delete and recreate ok?
+    // It certainly is simpler.
```

Comment:
FYI @daveverwer - I _think_ it'd be ok to just wipe and recreate versions (and later products) rather than try and match them up. It's just a lot less hassle than try and line up the adds and deletes with the futures but please let me know if there's a need to track the original creation date of versions. (As far as I can see the original creation date is the only bit of info we'd lose with this approach - in fact creation data would become pretty meaningless for these records.)

Reply:
I don't think the creation date of package version records is important at all, what's much more important is the version release date but that's not affected by this. So, I don't think there's any reason to need full reconciliation there.

I think I did reconciliation in the original implementation, there's something about the constant churning of integer database IDs that rubs my old-school brain the wrong way, but this uses UUID keys, so that's also not an issue.

Something about constantly deleting and re-inserting the same data (for versions that haven't changed) feels wrong to me, but there's no good reason apart from that.

Reply:
:+1: it's certainly something we could change going forward but I'll keep it simple at the start then!